### PR TITLE
Use non-refcounted tupdesc for multi-insert buffers

### PR DESCRIPTION
**Pull request by akuzm**
We don't need reference counting there, and it spends a lot of CPU in ResourceOwner.

It slightly improves COPY performance on some data sets:
```
benchmark:

environment_type: "timescaledb_singlenode"

prepare-benchmark:
  steps:
    - run: |
        drop table if exists uk_price_paid;
        create table uk_price_paid(price integer, "date" date, postcode1 text, postcode2 text, type smallint, is_new bool, duration smallint, addr1 text, addr2 text, street text, locality text, town text, district text, country text, category smallint);
        SELECT create_hypertable('uk_price_paid', 'date');
        COPY uk_price_paid FROM '/home/akuzm/dataset/uk-price-paid/prices-1M-2001-random-1.tsv';
        VACUUM ANALYZE;
        CHECKPOINT;

benchmark:
  executions: 5
  steps:
    - run: |
        BEGIN;
        COPY uk_price_paid FROM '/home/akuzm/dataset/uk-price-paid/prices-1M-2001-random-1.tsv';
        ROLLBACK;
      
    - run: |
        BEGIN;
        COPY uk_price_paid FROM '/home/akuzm/dataset/uk-price-paid/prices-1M-2001-ordered-1.tsv';
        ROLLBACK;
        
teardown:
  steps:
    - run: |
        DROP TABLE uk_price_paid;
        
        
main:
+-------------------------------------------------------------------------------------------+------------------------------------------+
| Query                                                                                     | 043bd55c0bf7ef372da6ad3e7874be030b10781c |
+-------------------------------------------------------------------------------------------+------------------------------------------+
| COPY uk_price_paid FROM '/home/akuzm/dataset/uk-price-paid/prices-1M-2001-random-1.tsv';  |                                  2266.51 |
| COPY uk_price_paid FROM '/home/akuzm/dataset/uk-price-paid/prices-1M-2001-ordered-1.tsv'; |                                  2409.68 |
+-------------------------------------------------------------------------------------------+------------------------------------------+

improved:
+-------------------------------------------------------------------------------------------+------------------------------------------+
| Query                                                                                     | 82de4f28bedb26caf18d7ff6c2241f268b6063ab |
+-------------------------------------------------------------------------------------------+------------------------------------------+
| COPY uk_price_paid FROM '/home/akuzm/dataset/uk-price-paid/prices-1M-2001-random-1.tsv';  |                                  2228.23 |
| COPY uk_price_paid FROM '/home/akuzm/dataset/uk-price-paid/prices-1M-2001-ordered-1.tsv'; |                                  2046.84 |
+-------------------------------------------------------------------------------------------+------------------------------------------+
```

### CODE REVIEW

File: src/copy.c [hunk 1]
```diff
@@ -193,6 +199,12 @@ TSCopyMultiInsertBufferInit(Point *point)
 	buffer->point = palloc(POINT_SIZE(point->num_coords));
 	memcpy(buffer->point, point, POINT_SIZE(point->num_coords));
 
+	/*
+	 * Make a non-refcounted copy of tupdesc to avoid spending CPU in
+	 * ResourceOwner when creating a big number of table slots.
+	 */
+	buffer->tupdesc = CreateTupleDescCopyConstr(cis->rel->rd_att);
```
Comment (jnidzwetzki) [inline src/copy.c:206]: Wouldn't it make sense to disable the reference counting for our local copy of the `tupledesc`? In my benchmarks, this results in a slight performance increase.

```suggestion
	buffer->tupdesc = CreateTupleDescCopyConstr(cis->rel->rd_att);
	/* Disable reference counting */
	buffer->tupdesc->tdrefcount = -1;
```

Comment (akuzm) [reply src/copy.c:206]: It should be already disabled for the copy, I'll add an assertion.

File: src/copy.c [hunk 2]
```diff
@@ -193,6 +199,12 @@ TSCopyMultiInsertBufferInit(Point *point)
 	buffer->point = palloc(POINT_SIZE(point->num_coords));
 	memcpy(buffer->point, point, POINT_SIZE(point->num_coords));
 
+	/*
+	 * Make a non-refcounted copy of tupdesc to avoid spending CPU in
+	 * ResourceOwner when creating a big number of table slots.
```
[the remark this answers is not in the document]
Comment (akuzm) [reply src/copy.c]: That's correct, added it to the comment.

File: src/copy.c (hunk 1 above)
Comment (jnidzwetzki) [reply src/copy.c:210]: Yes, you are right. I like the idea of adding an assertion to clearly indicate the disabled reference counting.

**Review by jnidzwetzki — APPROVED** [review]

**Review by erimatnor — APPROVED** [review]

**Pull request merged**